mesonbuild/mtest.py

```diff
@@ -946,6 +946,8 @@ def run_tests(self, tests: T.List['TestSerialisation']) -> None:
                             executor = conc.ThreadPoolExecutor(max_workers=self.options.num_processes)
                         f = executor.submit(single_test.run)
                         futures.append((f, numlen, tests, visible_name, i))
+                        if len(futures) > self.options.num_processes:
+                            self.drain_future(futures.pop(0))
```

Comment:
This has a negative impact on parallelism. If we have a long running test, this will block until that is finished but no new work can be started.

Reply:
That can easily be covered by adding an '* 2` on the number of queued process. But with current implementation, it means you'll have to wait double that time before it stops on error.

As I said in the message, I'm looking for feedback on what was the intent of this implementation. There is not enough implemented for me to guess what was expected here. Also, the strict ordering implemented here makes everything much more difficult in term of parallelism. So if you agree to drop that, we can likely just rewrite the entire thing with proper futures API.